networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/rpc/RpcHandler.java

```diff
@@ -58,53 +59,28 @@ public RpcHandler(final AsyncRunner asyncRunner, RpcMethod rpcMethod) {
   @SuppressWarnings("unchecked")
   public SafeFuture<RpcStream> sendRequest(
       Connection connection, Bytes initialPayload, RpcRequestHandler handler) {
-    if (connection.closeFuture().isDone()) {
-      return SafeFuture.failedFuture(new PeerDisconnectedException());
-    }
 
-    SafeFuture<RpcStream> streamFuture = new SafeFuture<>();
+    Interruptor closeInterruptor =
+        SafeFuture.createInterruptor(connection.closeFuture(), PeerDisconnectedException::new);
+    Interruptor timeoutInterruptor =
+        SafeFuture.createInterruptor(
+            asyncRunner.getDelayedFuture(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS),
+            () ->
+                new StreamTimeoutException(
+                    "Timed out waiting to initialize stream for method " + rpcMethod.getId()));
 
-    // Complete future if peer disconnects
-    SafeFuture.of(connection.closeFuture())
-        .always(() -> streamFuture.completeExceptionally(new PeerDisconnectedException()));
-    // Complete future if we fail to initialize
-    asyncRunner
-        .getDelayedFuture(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS)
-        .thenAccept(
-            __ ->
-                streamFuture.completeExceptionally(
-                    new StreamTimeoutException(
-                        "Timed out waiting to initialize stream for method " + rpcMethod.getId())))
-        .reportExceptions();
-
-    // Try to initiate stream
-    SafeFuture.of(
-            connection
-                .muxerSession()
-                .createStream(Multistream.create(this).toStreamHandler())
-                .getController())
+    return SafeFuture.notInterrupted(closeInterruptor)
         .thenCompose(
-            ctr -> {
-              ctr.setRequestHandler(handler);
-              return ctr.getRpcStream()
-                  .writeBytes(initialPayload)
-                  .thenApply(f -> ctr.getRpcStream())
-                  .thenAccept(
-                      rpcStream -> {
-                        if (!streamFuture.complete(rpcStream)) {
-                          // If future was already completed exceptionally, close the controller
-                          ctr.close();
-                        }
-                      });
-            })
-        .exceptionally(
-            err -> {
-              streamFuture.completeExceptionally(err);
-              return null;
-            })
-        .reportExceptions();
-
-    return streamFuture;
+            __ ->
+                connection
+                    .muxerSession()
+                    .createStream(Multistream.create(this).toStreamHandler())
+                    .getController())
+        .orInterrupt(closeInterruptor, timeoutInterruptor)
```

Comment:
The controller is anyway closed this or that way, but probably you are right it should be closed here explicitly. 
I'll get back to this as soon as sort out this NPE I posted below. It appeared a bit tricky

Reply:
Fixed this by closing the stream, since the controller could be still not created. The `Controller.close()` should anyway be invoked on `Stream.close`

Reply:
It seems pretty unlikely we'll actually hit this, but isn't it still possible for us to leave a dangling stream open here if the timeout interrupts before `getStream` resolves?

Reply:
It is possible in theory, but the `stream` is mostly created synchronously (when connection is active) so this situation seems pretty unlikely. In practice this situation can be assumed neglectable, though yes, strictly speaking in theory this is an issue.

Reply:
Would be nice to close this loophole somehow, but I'll leave it up to you if you think we'll never actually hit this case.